Add hotkey settings panel with a hotkey remap dialog

dragunoff opened the pull request:
Implement a hotkey settings remap dialog with validation (to prevent duplicate hotkeys) and buttons to unbind and reset to default. To make room for the dialog the hotkeys are moved to their own settings panel.

![изображение](https://user-images.githubusercontent.com/1355810/57957221-3d7dde80-7904-11e9-8a92-213b4b51a862.png)
The new hotkey panel before selecting a hotkey.

![изображение](https://user-images.githubusercontent.com/1355810/57956650-67ce9c80-7902-11e9-9a8b-86bc8d512eac.png)
The panel and dialog during rebinding (input is blinking)

![изображение](https://user-images.githubusercontent.com/1355810/57956869-183ca080-7903-11e9-9542-ca65d3f7f638.png)
Default binding

![изображение](https://user-images.githubusercontent.com/1355810/57956948-605bc300-7903-11e9-80d5-26765db5f86f.png)
Invalid binding (keeps blinking and won't save)

![изображение](https://user-images.githubusercontent.com/1355810/57957128-f09a0800-7903-11e9-8fb8-86a9650fb665.png)
Cleared binding (sets the key to UNKNOWN None)

![изображение](https://user-images.githubusercontent.com/1355810/57957383-c0069e00-7904-11e9-8a6f-27b4fc3b84fc.png)
I did it a little differently for TD (the buttons use glyphs).

This probably needs and upgrade rule somewhere. I'm very much open to feedback (code and UI/UX related).

Closes #16001
Closes #14670
Closes #3986
Closes #13200

dragunoff committed:
Scissor the text of a ButtonWidget if it overflows and display a tooltip

Also:
* add a variable to a comomn pattern used for truncating text in HotkeyEntryWidget and TextFieldWidget

dragunoff committed:
Add generic notice colors for the UI (info, success, warning, error)

dragunoff committed:
Remove from HotkeyManager hotkeys found in settings.yaml but not in hotkey definitons

dragunoff committed:
Add hotkey settings panel with a hotkey remap dialog

* Add HotkeyDialogLogic.cs
* Add dialog-hotkey.yaml to all mods
* Add `GetFirstDuplicate` method to HotkeyManager to aid in validation
* Add "Player" and/or "Spectator" type to all hotkeys to allow for
validation based on overlapping types
* Change settings.yaml and SettingsLogic.cs to work with the new dialog

dragunoff committed:
Change the name of Keycode.UNKNOWN to "Undefined"

dragunoff commented:
Updated and rebased.

Added initial validation for duplicates that may arise due to new hotkeys conflicting with user defined ones or the user manually editing settings.yaml.
![duplicate-hotkeys-initial](https://user-images.githubusercontent.com/1355810/58283983-5843cd80-7db2-11e9-8d13-35306d39cdd8.png)

My code in `SettingsLogic.cs` is starting to look like spaghetti...

teinarss commented on OpenRA.Mods.Common/Widgets/ButtonWidget.cs:
```diff
@@ -245,6 +245,14 @@ public override void Draw()
 			DrawBackground(rb, disabled, Depressed, hover, highlighted);
+
+			// Scissor when the text overflows
+			if (textSize.X > Bounds.Width - LeftMargin - RightMargin)
```
Wouldn't it be better to introduce a variable here? Because this is used twice.

dragunoff replied:
The same pattern is used in `HotkeyEntryWidget` and `TextFieldWidget`. Should I change it there as well?

dragunoff replied:
Fixed.

teinarss commented on OpenRA.Mods.Common/Widgets/Logic/HotkeyDialogLogic.cs:
```diff
@@ -0,0 +1,148 @@
+
+		void OnHotkeyEntryTakeFocus()
+		{
+			cancelButton.Disabled = manager.GetFirstDuplicate(definition.Name, currentHotkey, definition) != null ? true : false;
```
can be simplified to 
```suggestion
			cancelButton.Disabled = manager.GetFirstDuplicate(definition.Name, currentHotkey, definition) != null;
```

dragunoff replied:
Fixed.

teinarss commented:
I think so!

On Wed, Jun 19, 2019, 10:12 Ivaylo Draganov <notifications@github.com>
wrote:

> *@dragunoff* commented on this pull request.
> ------------------------------
>
> In OpenRA.Mods.Common/Widgets/ButtonWidget.cs
> <https://github.com/OpenRA/OpenRA/pull/16552#discussion_r295171522>:
>
> > @@ -245,6 +245,14 @@ public override void Draw()
>
>  			var hover = Ui.MouseOverWidget == this || Children.Any(c => c == Ui.MouseOverWidget);
>  			DrawBackground(rb, disabled, Depressed, hover, highlighted);
> +
> +			// Scissor when the text overflows
> +			if (textSize.X > Bounds.Width - LeftMargin - RightMargin)
>
> The same pattern is used in HotkeyEntryWidget and TextFieldWidget. Should
> I change it there as well?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/16552?email_source=notifications&email_token=AESI7GFRPWQ2EX62ZUEFEN3P3HTABA5CNFSM4HNYSQ62YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB37CPCA#discussion_r295171522>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AESI7GE3MFOH57YDY5WS2STP3HTABANCNFSM4HNYSQ6Q>
> .
>

dragunoff commented:
Rebased and updated to address comments.

While I was at it I also added a tootip for truncated button text:
![изображение](https://user-images.githubusercontent.com/1355810/59776855-327bec80-92bc-11e9-9d7b-1ea97d3a094a.png)

abcdefg30 commented:
Tbh, maybe we should consider increasing the width of those. It is not really obvious that `Ctrl + Shift` is truncated there.

dragunoff commented:
![изображение](https://user-images.githubusercontent.com/1355810/59835949-53dbe780-9353-11e9-8fee-52ae5731f834.png)

Added ellipsis to truncated buttons that have a tooltip container.

abcdefg30 commented:
Needs a rebase.

dragunoff commented:
Rebased.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/fd345eb5482f57dabb88d612c18daa2c2e6f9491...403b008162dc3e3546365a0142787744f431b2ca)

pchote commented on OpenRA.Mods.Common/Widgets/ButtonWidget.cs:
```diff
@@ -243,12 +243,27 @@ public override void Draw()
+			var isTextOverflowing = font.Measure(text).X > contentWidth;
+			if (isTextOverflowing && TooltipContainer != null)
+			{
+				Game.Renderer.EnableScissor(new Rectangle(RenderOrigin.X + LeftMargin, RenderOrigin.Y,
```
Why is this scissoring *and* truncating the text? The scissoring will be redundant here.

dragunoff replied:
You mean that truncating is sufficient? I didn't realize that... At first it was only scissoring (modeled after hotkey input/text input widget) but there had to be a visual cue so I added truncation. I can fix that in a follow up.

pchote replied:
`TruncateText` guarantees that the text width is less than `contentWidth`, provided that `contentWidth` is wide enough to fit at least "...".
